File: 0035. Search Insert Position/0035.cpp
//the easiest way to solve
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int right=nums.size()-1;
        int left=0;

        int middle;
        while(left<=right){
            middle=(left+right)/2;
            if (nums[middle]<target){
                left=middle + 1;
            }
            else if (nums[middle]>target)
                right=middle - 1;
            else{
                return middle;
            }
        }

        return left; // that is...
    }
};

File: 0035. Search Insert Position/0035.java
// recursion!
class Solution {
    public int searchInsertRecursion(int[] nums, int target, int left, int right){
        int middle=((right+left)/2);

        if (nums[middle]>target){
            // System.out.println("left");
            if (middle==0 || nums[middle-1]<target) return middle;
            return  searchInsertRecursion(nums, target, left, middle-1);
        }
        else if(nums[middle]==target){
            // System.out.println("middle");
            return middle;
        }
        else{
            // System.out.println("right");
            if (middle==nums.length-1 || nums[middle+1]>target) return middle+1;
            return searchInsertRecursion(nums, target, middle+1, right);
        }
    }
    
    public int searchInsert(int[] nums, int target) {
        return searchInsertRecursion(nums, target, 0, nums.length-1);
    }
}